book-content/chapters/09-typescript-only-features.md

```diff
@@ -0,0 +1,518 @@
+
+```typescript
+namespace RecordStoreUtils {
+  export function addAlbum(title: string, artist: string, year: number) {
+    // Implementation to add an album to the collection
+  }
+}
+```
+
+...would be transpiled into the following JavaScript:
+
+```javascript
+var RecordStoreUtils;
+(function (RecordStoreUtils) {
+  function addAlbum(title, artist, year) {
+    // Implementation to add an album to the collection
+  }
+  RecordStoreUtils.addAlbum = addAlbum;
+})(RecordStoreUtils || (RecordStoreUtils = {}));
+```
+
+Similarly to an enum, this code creates an object with properties for each function and type in the namespace. This means that the namespace can be accessed as an object, and its properties can be accessed as methods or properties.
+
+### Merging Namespaces
+
+Just like interfaces, namespaces can be merged through declaration merging. This allows you to combine two or more separate declarations into a single definition.
+
+Here we have two declarations of `RecordStoreUtils`– one with an `Album` namespace and another with a `Sales` namespace:
+
+```typescript
+namespace RecordStoreUtils {
+  export namespace Album {
+    export interface Album {
+      title: string;
+      artist: string;
+      year: number;
+    }
+  }
+}
+
+namespace RecordStoreUtils {
+  export namespace Sales {
+    export function recordSale(
+      albumTitle: string,
+      quantity: number,
+      price: number,
+    ) {
+      // Implementation to record an album sale
+    }
+
+    export function calculateTotalSales(albumTitle: string): number {
+      // Implementation to calculate total sales for an album
+      return 0; // Placeholder return
+    }
+  }
+}
+```
+
+Because namespaces support declaration merging, the two declarations are automatically combined into a single `RecordStoreUtils` namespace. Both the `Album` and `Sales` namespaces can be accessed as before:
+
+```typescript
+const loaded: RecordStoreUtils.Album.Album = {
+  title: "Loaded",
+  artist: "The Velvet Underground",
+  year: 1970,
+};
+
+RecordStoreUtils.Sales.calculateTotalSales("Loaded");
+```
+
+#### Merging Interfaces within Namespaces
+
+It's also possible for interfaces within namespaces to be merged. If we had two different `RecordStoreUtils` each with their own `Album` interface, TypeScript would automatically merge them into a single `Album` interface that includes all the properties:
+
+```typescript
+namespace RecordStoreUtils {
+  export interface Album {
+    title: string;
+    artist: string;
+    year: number;
+  }
+}
+
+namespace RecordStoreUtils {
+  export interface Album {
+    genre: string[];
+    recordLabel: string;
+  }
+}
+
+const madvillainy: RecordStoreUtils.Album = {
+  title: "Madvillainy",
+  artist: "Madvillain",
+  year: 2004,
+  genre: ["Hip Hop", "Experimental"],
+  recordLabel: "Stones Throw",
+};
+```
+
+This information will become crucial later when we look at the namespace's key use case: globally scoped types.
+
```

Comment:
I guess you could document this pattern too:
```ts
function test(opts: test.Options) {
  opts.foo;
}
namespace test {
  export type Options = {
    foo: string;
  };
}
```

what is nice about it that there is no added runtime to this, it's just a function with some types collocated to it